File: Android/scenes/convoai/src/main/java/io/agora/scene/convoai/ui/photo/PhotoProcessor.kt
package io.agora.scene.convoai.ui.photo

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import android.util.Log
import android.webkit.MimeTypeMap
import java.io.ByteArrayOutputStream

/**
 * Photo processor class - handles photo validation and processing
 * Processing flow:
 * 1. Format check -> 2. Resolution check -> 3. Size compression -> 4. Return processed photo
 */
object PhotoProcessor {
    
    private const val TAG = "PhotoProcessor"
    private const val MAX_DIMENSION = 2048 // Maximum width/height
    private const val MAX_FILE_SIZE = 5 * 1024 * 1024L // 5MB
    
    // Supported image formats
    private val SUPPORTED_FORMATS = setOf(
        "image/jpeg",
        "image/jpg", 
        "image/png",
        "image/webp"
    )
    
    /**
     * Process photo from URI
     * @param context Android context
     * @param uri Photo URI
     * @return Processed bitmap that meets requirements, or null if format not supported
     */
    fun processPhoto(context: Context, uri: Uri): Bitmap? {
        Log.d(TAG, "Starting photo processing for URI: $uri")
        
        try {
            // 1. Format check
            val mimeType = getMimeType(context, uri)
            Log.d(TAG, "Detected MIME type: $mimeType")
            
            if (mimeType == null || !SUPPORTED_FORMATS.contains(mimeType.lowercase())) {
                Log.w(TAG, "Unsupported image format: $mimeType")
                return null // Format not supported
            }
            
            // 用采样方式加载大图，防止内存暴涨
            val originalBitmap = loadBitmapFromUri(context, uri, MAX_DIMENSION, MAX_DIMENSION)
            if (originalBitmap == null) {
                Log.e(TAG, "Failed to load bitmap from URI")
                return null
            }
            
            return processBitmap(originalBitmap)
            
        } catch (e: Exception) {
            Log.e(TAG, "Error processing photo from URI: ${e.message}", e)
            return null
        }
    }
    
    /**
     * Process photo from Bitmap
     * @param bitmap Original bitmap
     * @return Processed bitmap that meets requirements, or null if processing fails
     */
    fun processPhoto(bitmap: Bitmap?): Bitmap? {
        if (bitmap == null) {
            Log.w(TAG, "Input bitmap is null")
            return null
        }
        
        return processBitmap(bitmap)
    }
    
    /**
     * Core bitmap processing logic
     */
    private fun processBitmap(originalBitmap: Bitmap): Bitmap? {
        try {
            Log.d(TAG, "Processing bitmap - Original size: ${originalBitmap.width}x${originalBitmap.height}")
            
            // 2. Resolution check and resize if needed
            val resizedBitmap = resizeIfNeeded(originalBitmap)
            Log.d(TAG, "After resize - Size: ${resizedBitmap.width}x${resizedBitmap.height}")
            
            // 3. Size compression if needed
            val finalBitmap = compressIfNeeded(resizedBitmap)
            Log.d(TAG, "After compression - Size: ${finalBitmap.width}x${finalBitmap.height}")
            
            Log.i(TAG, "Photo processing completed successfully")
            return finalBitmap
            
        } catch (e: Exception) {
            Log.e(TAG, "Error during bitmap processing: ${e.message}", e)
            return null
        }
    }
    
    /**
     * Resize image if dimensions exceed MAX_DIMENSION
     * Scale down the longer side to MAX_DIMENSION while maintaining aspect ratio
     */
    private fun resizeIfNeeded(bitmap: Bitmap): Bitmap {
        val originalWidth = bitmap.width.toFloat()
        val originalHeight = bitmap.height.toFloat()
        
        // Check if resize is needed
        if (originalWidth <= MAX_DIMENSION && originalHeight <= MAX_DIMENSION) {
            Log.d(TAG, "Image size is acceptable, no resize needed")
            return bitmap
        }
        
        // Calculate new size, maintaining aspect ratio
        val aspectRatio = originalWidth / originalHeight
        val newSize = if (originalWidth > originalHeight) {
            // Width is longer, scale based on width
            Pair(MAX_DIMENSION, (MAX_DIMENSION / aspectRatio).toInt())
        } else {
            // Height is longer, scale based on height
            Pair((MAX_DIMENSION * aspectRatio).toInt(), MAX_DIMENSION)
        }
        
        Log.d(TAG, "Resizing from ${originalWidth.toInt()}x${originalHeight.toInt()} to ${newSize.first}x${newSize.second}")
        
        return Bitmap.createScaledBitmap(bitmap, newSize.first, newSize.second, true)
    }
    
    /**
     * Compress image if estimated file size exceeds MAX_FILE_SIZE
     * Gradually reduce size until it meets the size requirement
     */
    private fun compressIfNeeded(bitmap: Bitmap): Bitmap {
        var currentBitmap = bitmap
        var estimatedSize = estimateJpegSize(currentBitmap, 80)
        
        Log.d(TAG, "Initial estimated JPEG size: ${formatFileSize(estimatedSize)}")
        
        // If size is within limit, return as is
        if (estimatedSize <= MAX_FILE_SIZE) {
            Log.d(TAG, "Image size is within limit, no compression needed")
            return currentBitmap
        }
        
        // Gradually reduce size until it meets the requirement
        var scaleFactor = 1.0f
        while (estimatedSize > MAX_FILE_SIZE && scaleFactor > 0.1f) {
            scaleFactor *= 0.8f // Reduce to 80% each iteration
            val newWidth = (bitmap.width * scaleFactor).toInt()
            val newHeight = (bitmap.height * scaleFactor).toInt()
            
            if (newWidth > 0 && newHeight > 0) {
                // Clean up previous bitmap if it's not the original
                if (currentBitmap != bitmap) {
                    currentBitmap.recycle()
                }
                
                currentBitmap = Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
                estimatedSize = estimateJpegSize(currentBitmap, 80)
                Log.d(TAG, "Compressed to: ${currentBitmap.width}x${currentBitmap.height}, estimated size: ${formatFileSize(estimatedSize)}")
            } else {
                break
            }
        }
        
        Log.i(TAG, "Final compressed size: ${currentBitmap.width}x${currentBitmap.height}, estimated JPEG size: ${formatFileSize(estimatedSize)}")
        return currentBitmap
    }
    
    /**
     * Estimate JPEG file size after compression
     */
    private fun estimateJpegSize(bitmap: Bitmap, quality: Int): Long {
        val pixels = bitmap.width * bitmap.height.toLong()
        val baseSize = pixels * 3 // RGB base size
        
        val compressionRatio = when {
            quality >= 90 -> 0.1f
            quality >= 80 -> 0.08f
            quality >= 70 -> 0.06f
            quality >= 60 -> 0.05f
            quality >= 50 -> 0.04f
            quality >= 40 -> 0.03f
            quality >= 30 -> 0.025f
            quality >= 20 -> 0.02f
            quality >= 10 -> 0.015f
            else -> 0.01f
        }
        
        return (baseSize * compressionRatio).toLong()
    }
    
    /**
     * Get file MIME type
     */
    private fun getMimeType(context: Context, uri: Uri): String? {
        return try {
            val contentResolver = context.contentResolver
            contentResolver.getType(uri) ?: run {
                // Fallback: get MIME type from file extension
                val fileExtension = MimeTypeMap.getFileExtensionFromUrl(uri.toString())
                if (fileExtension.isNotEmpty()) {
                    MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension.lowercase())
                } else {
                    null
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error getting MIME type: ${e.message}", e)
            null
        }
    }
    
    /**
     * Efficiently load large images with sampling to avoid memory issues
     */
    private fun loadBitmapFromUri(
        context: Context,
        uri: Uri,
        reqWidth: Int = MAX_DIMENSION,
        reqHeight: Int = MAX_DIMENSION
    ): Bitmap? {
        return try {
            val contentResolver = context.contentResolver

            // First decode: only get image dimensions
            val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            contentResolver.openInputStream(uri)?.use { inputStream ->
                BitmapFactory.decodeStream(inputStream, null, options)
            }

            // Calculate optimal inSampleSize
            options.inSampleSize = calculateInSampleSize(options, reqWidth, reqHeight)
            options.inJustDecodeBounds = false
            options.inPreferredConfig = Bitmap.Config.RGB_565

            // Second decode: load the actual bitmap with sampling
            contentResolver.openInputStream(uri)?.use { inputStream ->
                BitmapFactory.decodeStream(inputStream, null, options)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error loading bitmap from URI: ${e.message}", e)
            null
        }
    }

    /**
     * Calculate the optimal inSampleSize to ensure the loaded bitmap does not exceed the target size
     */
    private fun calculateInSampleSize(options: BitmapFactory.Options, reqWidth: Int, reqHeight: Int): Int {
        val (height: Int, width: Int) = options.outHeight to options.outWidth
        var inSampleSize = 1
        if (height > reqHeight || width > reqWidth) {
            val halfHeight: Int = height / 2
            val halfWidth: Int = width / 2
            while ((halfHeight / inSampleSize) >= reqHeight && (halfWidth / inSampleSize) >= reqWidth) {
                inSampleSize *= 2
            }
        }
        return inSampleSize
    }
    
    /**
     * Format file size for display
     */
    private fun formatFileSize(bytes: Long): String {
        return when {
            bytes < 1024 -> "${bytes}B"
            bytes < 1024 * 1024 -> "${bytes / 1024}KB"
            else -> String.format("%.1fMB", bytes / (1024.0 * 1024.0))
        }
    }
    
    /**
     * Rotate bitmap by specified degrees
     */
    fun rotateBitmap(bitmap: Bitmap, degrees: Float): Bitmap {
        val matrix = Matrix().apply {
            postRotate(degrees)
        }
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }
    
    /**
     * Flip bitmap horizontally (for front camera)
     */
    fun flipBitmap(bitmap: Bitmap): Bitmap {
        val matrix = Matrix().apply {
            postScale(-1f, 1f, bitmap.width / 2f, bitmap.height / 2f)
        }
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }
} 

File: iOS/Scenes/ConvoAI/ConvoAI/ConvoAI/Classes/Main/Photo/PhotoProcessor.swift
//
//  PhotoProcessor.swift
//  ConvoAI
//
//  Created by HeZhengQing on 2025/07/09.
//

import UIKit
import Foundation

/**
 * Photo processor class - handles photo validation and processing
 * Processing flow:
 * 1. Format check -> 2. Resolution check -> 3. Size compression -> 4. Return processed photo
 */
class PhotoProcessor {
    
    private static let TAG = "PhotoProcessor"
    private static let MAX_DIMENSION: CGFloat = 2048 // Maximum width/height
    private static let MAX_FILE_SIZE: Int64 = 5 * 1024 * 1024  // 5MB
    
    // Supported image formats
    private static let SUPPORTED_FORMATS: Set<String> = [
        "image/jpeg",
        "image/jpg",
        "image/png",
        "image/webp"
    ]
    
    /**
     * Process photo from URL
     * @param url Photo URL
     * @return Processed image that meets requirements, or nil if format not supported
     */
    static func processPhoto(url: URL) -> UIImage? {
        print("[\(TAG)] Starting photo processing for URL: \(url)")
        // 1. Format check
        let mimeType = getMimeType(from: url)
        print("[\(TAG)] Detected MIME type: \(mimeType ?? "unknown")")
        
        if let mimeType = mimeType {
            if !SUPPORTED_FORMATS.contains(mimeType.lowercased()) {
                print("[\(TAG)] Unsupported image format: \(mimeType)")
                return nil // Format not supported
            }
        } else {
            print("[\(TAG)] Could not determine MIME type")
            return nil
        }
        
        // Load image from URL
        guard let originalImage = UIImage(contentsOfFile: url.path) else {
            print("[\(TAG)] Failed to load image from URL")
            return nil
        }
        return processImage(originalImage)
    }
    
    /**
     * Process photo from UIImage
     * @param image Original image
     * @return Processed image that meets requirements, or nil if processing fails
     */
    static func processPhoto(_ image: UIImage?) -> UIImage? {
        guard let image = image else {
            print("[\(TAG)] Input image is nil")
            return nil
        }
        
        return processImage(image)
    }
    
    /**
     * Core image processing logic
     */
    private static func processImage(_ originalImage: UIImage) -> UIImage? {
        print("[\(TAG)] Processing image - Original size: \(originalImage.size.width)x\(originalImage.size.height)")
        
        // 2. Resolution check and resize if needed
        let resizedImage = resizeIfNeeded(originalImage)
        print("[\(TAG)] After resize - Size: \(resizedImage.size.width)x\(resizedImage.size.height)")
        
        // 3. Size compression if needed
        let finalImage = compressIfNeeded(resizedImage)
        print("[\(TAG)] After compression - Size: \(finalImage.size.width)x\(finalImage.size.height)")
        
        print("[\(TAG)] Photo processing completed successfully")
        return finalImage
    }
    
    /**
     * Resize image if dimensions exceed MAX_DIMENSION
     * Scale down the longer side to MAX_DIMENSION while maintaining aspect ratio
     */
    private static func resizeIfNeeded(_ image: UIImage) -> UIImage {
        let originalSize = image.size
        
        // Check if resize is needed
        if originalSize.width <= MAX_DIMENSION && originalSize.height <= MAX_DIMENSION {
            print("[\(TAG)] Image size is acceptable, no resize needed")
            return image
        }
        
        // Calculate new size, maintaining aspect ratio
        let aspectRatio = originalSize.width / originalSize.height
        var newSize: CGSize
        
        if originalSize.width > originalSize.height {
            // Width is longer, scale based on width
            newSize = CGSize(width: MAX_DIMENSION, height: MAX_DIMENSION / aspectRatio)
        } else {
            // Height is longer, scale based on height
            newSize = CGSize(width: MAX_DIMENSION * aspectRatio, height: MAX_DIMENSION)
        }
        
        print("[\(TAG)] Resizing from \(originalSize.width)x\(originalSize.height) to \(newSize.width)x\(newSize.height)")
        
        // Create graphics context and draw the resized image
        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        image.draw(in: CGRect(origin: .zero, size: newSize))
        let resizedImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return resizedImage ?? image
    }
    
    /**
     * Compress image if estimated file size exceeds MAX_FILE_SIZE
     * Gradually reduce size until it meets the size requirement
     */
    private static func compressIfNeeded(_ image: UIImage) -> UIImage {
        var currentImage = image
        var estimatedSize = estimateImageSize(currentImage)
        
        print("[\(TAG)] Initial estimated size: \(formatFileSize(estimatedSize))")
        
        // If size is within limit, return as is
        if estimatedSize <= MAX_FILE_SIZE {
            print("[\(TAG)] Image size is within limit, no compression needed")
            return currentImage
        }
        
        // Gradually reduce size until it meets the requirement
        var scaleFactor: CGFloat = 1.0
        while estimatedSize > MAX_FILE_SIZE && scaleFactor > 0.1 {
            scaleFactor *= 0.8 // Reduce to 80% each iteration
            let newSize = CGSize(
                width: image.size.width * scaleFactor,
                height: image.size.height * scaleFactor
            )
            
            if newSize.width > 0 && newSize.height > 0 {
                UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
                image.draw(in: CGRect(origin: .zero, size: newSize))
                if let scaledImage = UIGraphicsGetImageFromCurrentImageContext() {
                    currentImage = scaledImage
                    estimatedSize = estimateImageSize(currentImage)
                    print("[\(TAG)] Compressed to: \(currentImage.size.width)x\(currentImage.size.height), estimated size: \(formatFileSize(estimatedSize))")
                }
                UIGraphicsEndImageContext()
            } else {
                break
            }
        }
        
        print("[\(TAG)] Final compressed size: \(currentImage.size.width)x\(currentImage.size.height), estimated size: \(formatFileSize(estimatedSize))")
        return currentImage
    }
    
    /**
     * Estimate the file size of UIImage
     */
    private static func estimateImageSize(_ image: UIImage) -> Int64 {
        let width = image.size.width
        let height = image.size.height
        let scale = image.scale
        
        // Calculate actual pixel count
        let pixelWidth = width * scale
        let pixelHeight = height * scale
        let totalPixels = Int64(pixelWidth * pixelHeight)
        
        // Estimate bytes based on image characteristics (JPEG compression ratio is about 5-15% of original RGBA)
        let bytesPerPixel: Int64 = 4 // RGBA
        let rawSize = totalPixels * bytesPerPixel
        let compressionRatio: Double = 0.1 // 10% compression ratio, conservative estimate
        
        let estimatedSize = Int64(Double(rawSize) * compressionRatio)
        return estimatedSize
    }
    
    /**
     * Get file MIME type
     */
    private static func getMimeType(from url: URL) -> String? {
        let pathExtension = url.pathExtension.lowercased()
        
        switch pathExtension {
        case "jpg", "jpeg":
            return "image/jpeg"
        case "png":
            return "image/png"
        case "webp":
            return "image/webp"
        default:
            return nil
        }
    }
    
    /**
     * Format file size display
     */
    private static func formatFileSize(_ bytes: Int64) -> String {
        if bytes < 1024 {
            return "\(bytes)B"
        } else if bytes < 1024 * 1024 {
            return "\(bytes / 1024)KB"
        } else {
            return String(format: "%.1fMB", Double(bytes) / (1024.0 * 1024.0))
        }
    }
    
    /**
     * Rotate image by specified degrees
     */
    static func rotateImage(_ image: UIImage, degrees: CGFloat) -> UIImage {
        let radians = degrees * .pi / 180
        let rotatedSize = CGRect(origin: .zero, size: image.size)
            .applying(CGAffineTransform(rotationAngle: radians))
            .integral.size
        
        UIGraphicsBeginImageContextWithOptions(rotatedSize, false, image.scale)
        if let context = UIGraphicsGetCurrentContext() {
            let origin = CGPoint(x: rotatedSize.width / 2, y: rotatedSize.height / 2)
            context.translateBy(x: origin.x, y: origin.y)
            context.rotate(by: radians)
            image.draw(in: CGRect(
                x: -image.size.width / 2,
                y: -image.size.height / 2,
                width: image.size.width,
                height: image.size.height
            ))
        }
        let rotatedImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return rotatedImage ?? image
    }
    
    /**
     * Flip image horizontally (for front camera)
     */
    static func flipImage(_ image: UIImage) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
        if let context = UIGraphicsGetCurrentContext() {
            context.translateBy(x: image.size.width, y: 0)
            context.scaleBy(x: -1, y: 1)
            image.draw(in: CGRect(origin: .zero, size: image.size))
        }
        let flippedImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return flippedImage ?? image
    }
} 
